src/gwt/src/org/rstudio/studio/client/workbench/views/terminal/TerminalSessionSocket.java

```diff
@@ -309,15 +326,42 @@ public void dispatchInput(int inputSequence,
       default:
          break;
       }
-      
    }
    
    /**
     * Send output to the terminal emulator.
     * @param output text to send to the terminal
+    * @param detectLocalEcho local-echo detection
     */
-   public void dispatchOutput(String output)
+   public void dispatchOutput(String output, boolean detectLocalEcho)
    {
+      if (detectLocalEcho && !localEcho_.isEmpty())
+      {
+         String lastOutput = "";
+         while (!localEcho_.isEmpty() && lastOutput.length() < output.length())
```

Comment:
I have a feeling we'll need to, but certainly worth taking this as is now and seeing what shakes out. Thanks for those improvements, this is going to make a huge difference in the responsiveness!